C/STrace/EtwLogger.cpp

```diff
@@ -0,0 +1,460 @@
+#include "EtwLogger.h"
+
+#include "vector.h"
+
+#define va_copy(dest, src) (dest = src)
+
+// Cache of all created providers.
+MyVector<detail::EtwProvider> g_ProviderCache;
+
+typedef enum _ETW_FIELD_TYPE
+{
+	EtwFieldNull,
+	EtwFieldUnicodeString,
+	EtwFieldAnsiString,
+	EtwFieldInt8,
+	EtwFieldUInt8,
+	EtwFieldInt16,
+	EtwFieldUInt16,
+	EtwFieldInt32,
+	EtwFieldUInt32,
+	EtwFieldInt64,
+	EtwFieldUInt64,
+	EtwFieldFloat,
+	EtwFieldDouble,
+	EtwFieldBool32,
+	EtwFieldBinary,
+	EtwFieldGuid,
+	EtwFieldPointer,
+	EtwFieldFiletime,
+	EtwFieldSystemTime,
+	EtwFieldSid,
+	EtwFieldHexInt32,
+	EtwFieldHexInt64,
+	EtwFieldPid = (EtwFieldInt32 | 0x05 << 8),
+} ETW_FIELD_TYPE;
+
+namespace detail
+{
+
+#pragma region EtwProvider
+
+EtwProvider::EtwProvider(LPCGUID providerGuid) : m_guid{ providerGuid }, m_regHandle{}, m_providerMetadataDesc{}, m_events{}
+{
+}
+
+EtwProvider::EtwProvider(EtwProvider&& other)
+{
+	m_guid = move(other.m_guid);
+	m_regHandle = move(other.m_regHandle);
+	m_providerMetadataDesc = move(other.m_providerMetadataDesc);
+	m_events = move(other.m_events);
+}
+
+EtwProvider& EtwProvider::operator=(EtwProvider&& other)
+{
+	m_guid = move(other.m_guid);
+	m_regHandle = move(other.m_regHandle);
+	m_providerMetadataDesc = move(other.m_providerMetadataDesc);
+	m_events = move(other.m_events);
+
+	return *this;
+}
+
+NTSTATUS EtwProvider::Initialize(const char* providerName)
+{
+	// Register the kernel-mode ETW provider.
+	auto status = EtwRegister(m_guid, NULL, NULL, &m_regHandle);
+	if (status != STATUS_SUCCESS)
+	{
+		return status;
+	}
+
+	// Create packaged provider metadata structure.
+	// <https://learn.microsoft.com/en-us/windows/win32/etw/provider-traits>
+	const auto providerMetadataLength = (uint16_t)((strlen(providerName) + 1) + sizeof(uint16_t));
+	const auto providerMetadata = (struct ProviderMetadata*)ExAllocatePoolWithTag(NonPagedPoolNx, providerMetadataLength, 'wteO');
+	if (providerMetadata == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+
+	memset(providerMetadata, 0, providerMetadataLength);
+	providerMetadata->TotalLength = providerMetadataLength;
+	strcpy(providerMetadata->ProviderName, providerName);
+
+	// Tell the provider to use the metadata structure.
+	status = EtwSetInformation(m_regHandle, EventProviderSetTraits, providerMetadata, providerMetadataLength);
+	if (status != STATUS_SUCCESS)
+	{
+		return status;
+	}
+
+	// Create an EVENT_DATA_DESCRIPTOR pointing to the metadata.
+	EventDataDescCreate(&m_providerMetadataDesc, providerMetadata, providerMetadata->TotalLength);
+	m_providerMetadataDesc.Type = EVENT_DATA_DESCRIPTOR_TYPE_PROVIDER_METADATA;  // Descriptor contains provider metadata.
+
+	return STATUS_SUCCESS;
+}
+
+void EtwProvider::Destruct()
+{
+	for (auto i = 0; i < m_events.len(); i++)
```

Comment:
Not necessary to call this yourself, you can use the vector variant that invokes a real destructor here, it will free itself and it's contained object. You really want two type of `MyVector` it seems like, one that does not itself have a destructor/constructor and one that does. In both cases the `MyVector` will call the c++ destructor of its contained children and free those automatically.